src/main/java/org/apache/spark/shuffle/ucx/memory/MemoryPool.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (C) Mellanox Technologies Ltd. 2019. ALL RIGHTS RESERVED.
+ * See file LICENSE for terms.
+ */
+package org.apache.spark.shuffle.ucx.memory;
+
+import org.apache.spark.shuffle.UcxShuffleConf;
+import org.apache.spark.unsafe.Platform;
+import org.openucx.jucx.ucp.UcpContext;
+import org.openucx.jucx.ucp.UcpMemory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.nio.ByteBuffer;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Utility class to reuse and preallocate registered memory to avoid memory allocation
+ * and registration during shuffle phase.
+ */
+public class MemoryPool implements Closeable {
+  private static final Logger logger = LoggerFactory.getLogger(MemoryPool.class);
+
+  @Override
+  public void close() {
+    for (AllocatorStack stack: allocStackMap.values()) {
+      stack.close();
+      logger.info("Stack of size {}. " +
+          "Total requests: {}, total allocations: {}, preAllocations: {}",
+        stack.length, stack.totalRequests.get(), stack.totalAlloc.get(), stack.preAllocs.get());
+    }
+    allocStackMap.clear();
+  }
+
+  private class AllocatorStack implements Closeable {
+    private final AtomicInteger totalRequests = new AtomicInteger(0);
+    private final AtomicInteger totalAlloc = new AtomicInteger(0);
+    private final AtomicInteger preAllocs = new AtomicInteger(0);
+    private final ConcurrentLinkedDeque<RegisteredMemory> stack = new ConcurrentLinkedDeque<>();
+    private final int length;
+
+    private AllocatorStack(int length) {
+      this.length = length;
+    }
+
+    private RegisteredMemory get() {
+      RegisteredMemory result = stack.pollFirst();
+      if (result == null) {
+        if (length < conf.minRegistrationSize()) {
+          int numBuffers = conf.minRegistrationSize() / length;
+          logger.debug("Allocating {} buffers of size {}", numBuffers, length);
+          preallocate(numBuffers);
+          return get();
```

Comment:
Highly unlikely but theoretically possible that all preallocation buffers got consumed by other threads, so `stack.pollFirst()` would return null

Reply:
Or even in a situation where `maxRegisterSize = 4mb` - requested buffer of size 3.9 mb, it preallocated 1 buffer - other thread got it.

Reply:
With a recursive call, it'll again preallocate and try to get a buffer.

Reply:
1. isn't the length in this case always power of 2 ("private final int length")
2. maybe return one buffer from preallocate() to the caller to avoid threads stealing other threads' buffers?

Reply:
1. The length is final for the inner class of AllocatorStack. It works the follows. There's a map of size -> allocatorStack. When requested a buffer of size - it gets from the map needed AllocatorStack or creates a new one:
```java
AllocatorStack stack =
      allocStackMap.computeIfAbsent((int)roundedSize, AllocatorStack::new);
```
2. Preallocate puts buffer to the stack, to keep the same logic when preallocation happens beforehand.

Reply:
so, maybe instead of recursion it could be iteration? e.g replace 'if (result==null)' by 'while (result==null)'
i think recursion will have more overhead compared to iteration

Reply:
Changed to recurse only in rare cases when stack.poll retruned null